crates/boojum/src/gadgets/tower_extension/algebraic_torus.rs

```diff
@@ -0,0 +1,494 @@
+    {
+        // Intializing the result with 1
+        let mut result = Self::one(cs, self.get_params());
+
+        // Preparing self and self inverse in advance
+        let mut self_cloned = self.clone();
+        let mut self_inverse = self.conjugate(cs);
+
+        for bit in decomposition.as_ref().iter() {
+            result = result.square(cs);
+
+            // If bit is 1, multiply by initial torus
+            let bit_is_one = Boolean::allocated_constant(cs, *bit == 1);
+            let result_times_self = result.mul(cs, &mut self_cloned);
+            result = Self::conditionally_select(cs, bit_is_one, &result_times_self, &result);
+
+            // If bit is -1, multiply by inverse initial torus
+            let bit_is_minus_one = Boolean::allocated_constant(cs, *bit == -1);
+            let result_times_self_inverse = result.mul(cs, &mut self_inverse);
+            result = Self::conditionally_select(
+                cs,
+                bit_is_minus_one,
+                &result_times_self_inverse,
+                &result,
+            );
+        }
+
+        result
+    }
+
+    pub fn pow_u32<CS, S: AsRef<[u64]>>(&mut self, cs: &mut CS, exponent: S) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut result = Self::one(cs, self.get_params());
+        let mut found_one = false;
+
+        for bit in BitIterator::new(exponent) {
+            let apply_squaring = Boolean::allocated_constant(cs, found_one);
+            let result_squared = result.square(cs);
+            result = Self::conditionally_select(cs, apply_squaring, &result_squared, &result);
+            if !found_one {
+                found_one = bit;
+            }
+
+            let result_multiplied = result.mul(cs, self);
+            let apply_multiplication = Boolean::allocated_constant(cs, bit);
+            result =
+                Self::conditionally_select(cs, apply_multiplication, &result_multiplied, &result);
+
+            result.normalize(cs);
+        }
+
+        result
+    }
+
+    pub fn square<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // We compute squaring unconstrained:
+        let witness = self.encoding_to_witness(cs);
+        let witness_squared = P::torus_square(witness);
+
+        // Now, we constraint squaring with a cheaper version:
+        // g' = (1/2)(g + \gamma/g) is equivalent to
+        // (2g' - g)*g = gamma
+        let params = self.encoding.get_params();
+        let encoding_new = Fq6::allocate_from_witness(cs, witness_squared, params);
+
+        // lhs = (2g' - g)*g
+        let mut lhs = encoding_new.clone();
+        lhs = lhs.double(cs);
+        lhs = lhs.sub(cs, &mut self.encoding.clone());
+        let lhs = self.encoding.clone().mul(cs, &mut lhs);
+
+        // rhs = g == 0 ? zero : gamma
+        let zero = Fq6::zero(cs, params);
+        let gamma = Fq6::gamma(cs, params);
+        let is_zero_g = self.encoding.is_zero(cs);
```

Comment:
The same is mention here

Reply:
I mean, it isn't possible that gamma could be zero

Reply:
In case of multiplication, we ensure that `g + g'` cannot be zero due to algebraic impossibility of that. Here, on the other hand, it is completely possible that `g` will be zero, leading to division by that. That's why we need the check; I mean, in code, the `gamma/g` must be masked to zero, if g is zero.

This is particularly what is happening in `franklin-crypto`: 

https://github.com/matter-labs/franklin-crypto/blob/finalization/src/plonk/circuit/bigint/algebraic_torus.rs#L254